onnxruntime/core/providers/cpu/generator/random.cc

```diff
@@ -84,15 +84,15 @@ static TensorProto::DataType InferDataType(const Tensor& tensor);
 Status RandomNormal::Compute(OpKernelContext* ctx) const {
   Tensor& Y = *ctx->Output(0, shape_);
 
-  auto status = RandomNormalCompute(mean_, scale_, generator_, dtype_, Y);
+  auto status = RandomNormalCompute(mean_, scale_, const_cast<std::default_random_engine&>(generator_), dtype_, Y);
```

Comment:
>const_cast [](start = 51, length = 10)

const_cast is a massive code smell when used like this. Compute can be called concurrently, so sharing a generator across that leads to non-deterministic behaviour. I know we're generating random numbers, but are they expected/required to be deterministic based on the seed?

Reply:
I totally agree regarding the const_cast. Probably there is a better fix that still takes into consideration of the immutability of provider kernels.
The issue with current ORT:
load a model with random generator op and do inference with input-data1, it gets output-data1. with same loaded model, do inference again with the same input-data1,  it gets the same output-data1 again. This is NOT the expected/desired behavior of the model. The desired behavior is: with the same input, repeat evaluations shall produce similar but different outputs.

Reply:
An alternative option to mitigate the non-deterministic behaviour a little would be keeping the original code and passing back the copied generator from the GenerateData call. That could be used to update the generator_ instance in the operator (and make generator_ mutable and add a big comment explaining why instead of using const_cast). This way each GenerateData is using its own instance so we're not interleaving calls to a single generator instance from different GenerateData calls.

---
In reply to: [256618677](https://github.com/Microsoft/onnxruntime/pull/477#discussion_r256618677) [](ancestors = 256618677)